Add FlexAttention Op to ai.onnx.preview opset

mshr-h opened the pull request:
### Description

This PR adds the FlexAttention operator to the ai.onnx.preview domain.

FlexAttention mirrors the capabilities of [PyTorch's torch.nn.functional.flex_attention](https://docs.pytorch.org/docs/stable/nn.attention.flex_attention.html), enabling highly customizable attention mechanisms through user-defined subgraphs. This will unblock the export of models from libraries like Hugging Face Transformers that use attn_implementation="flex_attention" with [AttentionInterface](https://huggingface.co/docs/transformers/en/attention_interface).

### Motivation and Context

Discussed in #7494 

Design considerations
- Support rank-4 input tensors (batch, heads, seq_len, head_dim) only, matching PyTorch's FlexAttention interface.
- The function body is implemented using Loop ops to apply modifier subgraphs element-wise. While this is not performant, it serves as a correct reference implementation that works on any ONNX backend.

I validated the reference evaluator and the function body against PyTorch's FlexAttention using the following testcode. All tests pass with rtol=1.3e-6, atol=1e-5 against PyTorch reference. It's not included in the PR.

<details>

<summary> flexattn-test.py </summary>

```python
from typing import Callable

import numpy as np
import onnx.inliner
import pytest
import torch
from onnx.reference import ReferenceEvaluator
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
from torch.onnx import ONNXProgram

import onnx
from onnx import TensorProto, helper

torch.nn.attention.flex_attention._FLEX_ATTENTION_DISABLE_COMPILE_DEBUG = True  # ty:ignore[possibly-missing-attribute]


class ScoreModONNXWrapper(torch.nn.Module):
    """Wraps PyTorch score_mod (+ optional mask_mod) into a single-input
    ONNX-compatible module: scores[B,Hq,L,S] -> scores_out[B,Hq,L,S]."""

    def __init__(
        self,
        score_mod_func: Callable | None = None,
        mask_mod_func: Callable | None = None,
        mask_value: float = -float("inf"),
    ):
        super().__init__()
        self.score_mod_func = score_mod_func  # ty:ignore[unresolved-attribute]
        self.mask_mod_func = mask_mod_func  # ty:ignore[unresolved-attribute]
        self.mask_value = mask_value  # ty:ignore[unresolved-attribute]

    def forward(self, scores: torch.Tensor) -> torch.Tensor:
        B, Hq, L, S = scores.shape

        # Build broadcast-compatible index tensors
        b = torch.arange(B, device=scores.device, dtype=torch.int64).reshape(B, 1, 1, 1)
        h = torch.arange(Hq, device=scores.device, dtype=torch.int64).reshape(1, Hq, 1, 1)
        q_idx = torch.arange(L, device=scores.device, dtype=torch.int64).reshape(1, 1, L, 1)
        k_idx = torch.arange(S, device=scores.device, dtype=torch.int64).reshape(1, 1, 1, S)

        # Apply score_mod (element-wise with broadcasting)
        if self.score_mod_func is not None:
            scores = self.score_mod_func(scores, b, h, q_idx, k_idx)

        # Fold mask_mod into score_mod via Where
        if self.mask_mod_func is not None:
            mask = self.mask_mod_func(b, h, q_idx, k_idx)
            scores = torch.where(mask, scores, torch.tensor(self.mask_value, dtype=scores.dtype))

        return scores


def make_score_mod_graph(
    score_mod_func: Callable | None = None,
    mask_mod_func: Callable | None = None,
    B: int = 2,
    Hq: int = 2,
    L: int = 4,
    S: int = 4,
    mask_value: float = -float("inf"),
) -> onnx.GraphProto | None:
    """Auto-construct an ONNX score_mod subgraph via torch.onnx.export.

    Returns None if both score_mod_func and mask_mod_func are None.
    """
    if score_mod_func is None and mask_mod_func is None:
        return None

    wrapper = ScoreModONNXWrapper(score_mod_func, mask_mod_func, mask_value).eval()
    dummy_scores = torch.zeros(B, Hq, L, S, dtype=torch.float32)

    onnx_program: ONNXProgram = torch.onnx.export(
        wrapper,
        (dummy_scores,),
        input_names=["scores"],
        output_names=["scores_out"],
        opset_version=18,
        verbose=False,
    )  # ty:ignore[invalid-assignment]
    return onnx_program.model_proto.graph


def make_flexattention_onnx(
    Q: torch.Tensor,
    K: torch.Tensor,
    V: torch.Tensor,
    scale: float | None = None,
    score_mod_torch: Callable | None = None,
    mask_mod_torch: Callable | None = None,
) -> onnx.ModelProto:
    """Build a schema-compliant FlexAttention ONNX model.

    - score_mod and mask_mod are fused into a single score_mod graph attribute.
    - No mask_mod / enable_gqa / mask_value attributes (not in schema).
    - GQA is inferred automatically from Hq vs Hkv.
    """
    B, Hq, L, E = Q.shape
    _Bk, Hkv, S, _Ek = K.shape
    _Bv, _Hkv2, _Sv, Ev = V.shape

    q_in = helper.make_tensor_value_info("query", TensorProto.FLOAT, list(Q.shape))
    k_in = helper.make_tensor_value_info("key", TensorProto.FLOAT, list(K.shape))
    v_in = helper.make_tensor_value_info("value", TensorProto.FLOAT, list(V.shape))
    y_out = helper.make_tensor_value_info("output", TensorProto.FLOAT, [B, Hq, L, Ev])

    # Build score_mod graph (fusing mask_mod if present)
    score_mod_graph = make_score_mod_graph(
        score_mod_func=score_mod_torch,
        mask_mod_func=mask_mod_torch,
        B=B,
        Hq=Hq,
        L=L,
        S=S,
    )

    # Build node kwargs (only include attributes that are set)
    node_kwargs = {}
    if scale is not None:
        node_kwargs["scale"] = scale
    if score_mod_graph is not None:
        node_kwargs["score_mod"] = score_mod_graph

    node = helper.make_node(
        op_type="FlexAttention",
        inputs=["query", "key", "value"],
        outputs=["output"],
        domain="ai.onnx.preview",
        **node_kwargs,
    )

    graph = helper.make_graph(
        nodes=[node],
        name="flexattn_graph",
        inputs=[q_in, k_in, v_in],
        outputs=[y_out],
    )

    model = helper.make_model(
        graph,
        opset_imports=[
            helper.make_opsetid("", 18),
            helper.make_opsetid("ai.onnx.preview", 1),
        ],
        producer_name="flexattn_minimum_test",
    )
    onnx.checker.check_model(model)
    return model


def print_diff_stats(a: np.ndarray, b: np.ndarray) -> None:
    diff = a.astype(np.float64) - b.astype(np.float64)
    max_abs = float(np.max(np.abs(diff)))
    den = np.maximum(np.abs(a.astype(np.float64)), 1e-12)
    max_rel = float(np.max(np.abs(diff) / den))
    print("  max_abs_diff:", max_abs)
    print("  max_rel_diff:", max_rel)


# ---------------------------------------------------------------------------
# 3-way validation: PyTorch vs ONNX (original) vs ONNX (inlined)
# ---------------------------------------------------------------------------


def validate_flexattention(
    score_mod: Callable | None,
    mask_mod: Callable | None,
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
    scale: float | None,
) -> None:
    q_np = q.cpu().numpy()
    k_np = k.cpu().numpy()
    v_np = v.cpu().numpy()

    B, Hq, L, _E = q.shape
    _Bk, Hkv, S, _Ek = k.shape

    # --- (1) PyTorch reference ---
    enable_gqa = Hq != Hkv
    with torch.no_grad():
        block_mask = None
        if mask_mod is not None:
            block_mask = create_block_mask(mask_mod, B, Hq, L, S, device=q.device)

        y_pt = flex_attention(
            q,
            k,
            v,
            score_mod=score_mod,
            block_mask=block_mask,
            scale=scale,
            enable_gqa=enable_gqa,
        )
        if isinstance(y_pt, (tuple, list)):
            y_pt = y_pt[0]
        y_pt_np = y_pt.cpu().numpy()

    # --- (2) ONNX original model ---
    onnx_model = make_flexattention_onnx(
        q,
        k,
        v,
        scale=scale,
        score_mod_torch=score_mod,
        mask_mod_torch=mask_mod,
    )
    evaluator = ReferenceEvaluator(onnx_model)
    y_onnx = evaluator.run(None, {"query": q_np, "key": k_np, "value": v_np})[0]  # ty:ignore[invalid-argument-type]

    rtol, atol = 1.3e-6, 1e-5
    ok = torch.allclose(y_pt.cpu(), torch.from_numpy(y_onnx), rtol=rtol, atol=atol)

    print()
    print(f"[original] torch.allclose(rtol={rtol}, atol={atol}): {'✅' if ok else '❌'}")
    print_diff_stats(y_pt_np, y_onnx)
    assert ok, "ONNX FlexAttention (original) does not match PyTorch FlexAttention"

    # --- (3) ONNX inlined model ---
    inlined_onnx_model = onnx.inliner.inline_selected_functions(
        onnx_model,
        [("ai.onnx.preview", "FlexAttention")],
        exclude=False,
        inline_schema_functions=True,
    )
    onnx.checker.check_model(inlined_onnx_model)

    flexattn_nodes = [n for n in inlined_onnx_model.graph.node if n.op_type == "FlexAttention"]
    assert len(flexattn_nodes) == 0, f"Expected 0 FlexAttention nodes after inlining, got {len(flexattn_nodes)}"

    inlined_evaluator = ReferenceEvaluator(inlined_onnx_model)
    y_onnx_inlined = inlined_evaluator.run(  # ty:ignore[invalid-argument-type]
        None, {"query": q_np, "key": k_np, "value": v_np}
    )[0]

    inlined_ok = torch.allclose(y_pt.cpu(), torch.from_numpy(y_onnx_inlined), rtol=rtol, atol=atol)
    print(f"[inlined]  torch.allclose(rtol={rtol}, atol={atol}): {'✅' if inlined_ok else '❌'}")
    print_diff_stats(y_pt_np, y_onnx_inlined)
    assert inlined_ok, "ONNX FlexAttention (inlined) does not match PyTorch FlexAttention"


# ---------------------------------------------------------------------------
# PyTorch score_mod functions
# ---------------------------------------------------------------------------


def noop(score, b, h, q_idx, k_idx):
    return score


def relative_positional(score, b, h, q_idx, k_idx):
    return score + (q_idx - k_idx)


def alibi_causal(score, b, h, q_idx, k_idx):
    alpha = -0.1
    mask_const = -1e9
    dq = q_idx.to(dtype=score.dtype)
    dk = k_idx.to(dtype=score.dtype)
    biased = score + (dq - dk) * score.new_tensor(alpha)
    allowed = k_idx <= q_idx
    return torch.where(allowed, biased, score.new_tensor(mask_const))


def soft_cap(score, b, h, q_idx, k_idx):
    softcap = 20
    score = score / softcap
    score = torch.tanh(score)
    score = score * softcap
    return score


def causal_score(score, b, h, q_idx, k_idx):
    return torch.where(q_idx >= k_idx, score, -float("inf"))


# ---------------------------------------------------------------------------
# PyTorch mask_mod functions
# ---------------------------------------------------------------------------


def causal_mask(b: torch.Tensor, h: torch.Tensor, q_idx: torch.Tensor, k_idx: torch.Tensor) -> torch.Tensor:
    return k_idx <= q_idx


def sliding_window_causal(b, h, q_idx, k_idx):
    SLIDING_WINDOW = 2
    causal = q_idx >= k_idx
    window = q_idx - k_idx <= SLIDING_WINDOW
    return causal & window


# ---------------------------------------------------------------------------
# Test fixtures and parametrization
# ---------------------------------------------------------------------------


@pytest.fixture(scope="module")
def random_tensors():
    rng = np.random.default_rng(42)
    torch.manual_seed(42)
    B, H, L, S, E, Ev = 2, 2, 4, 4, 8, 8
    q_np = rng.standard_normal((B, H, L, E)).astype(np.float32)
    k_np = rng.standard_normal((B, H, S, E)).astype(np.float32)
    v_np = rng.standard_normal((B, H, S, Ev)).astype(np.float32)
    return torch.from_numpy(q_np), torch.from_numpy(k_np), torch.from_numpy(v_np)


@pytest.fixture(scope="module")
def random_tensors_gqa():
    """True GQA: Hq > Hkv and Hq % Hkv == 0."""
    rng = np.random.default_rng(43)
    torch.manual_seed(43)
    B, Hq, Hkv, L, S, E, Ev = 2, 8, 2, 4, 4, 8, 8
    q_np = rng.standard_normal((B, Hq, L, E)).astype(np.float32)
    k_np = rng.standard_normal((B, Hkv, S, E)).astype(np.float32)
    v_np = rng.standard_normal((B, Hkv, S, Ev)).astype(np.float32)
    return torch.from_numpy(q_np), torch.from_numpy(k_np), torch.from_numpy(v_np)


# fmt: off
# Each entry: (score_mod_name, score_mod_func, mask_mod_name, mask_mod_func, use_gqa, scale)
CASES: list[tuple[str, Callable | None, str, Callable | None, bool, float | None]] = [
    # --- Basic cases ---
    ("none"        , None               , "none"                 , None                 , False, None),
    ("noop"        , noop               , "none"                 , None                 , False, 1.0),
    ("none"        , None               , "causal_mask"          , causal_mask          , False, 1.0),
    ("relative_pos", relative_positional, "none"                 , None                 , False, 1.0),
    ("soft_cap"    , soft_cap           , "none"                 , None                 , False, 1.0),
    ("alibi_causal", alibi_causal       , "none"                 , None                 , False, 1.0),
    ("none"        , None               , "sliding_window_causal", sliding_window_causal, False, 1.0),
    # --- Composite: score_mod + mask_mod ---
    ("relative_pos", relative_positional, "causal_mask"          , causal_mask          , False, 1.0),
    # --- GQA ---
    ("none"        , None               , "none"                 , None                 , True , 1.0),
    ("none"        , None               , "causal_mask"          , causal_mask          , True , 1.0),
    # --- Scale variations ---
    ("none"        , None               , "causal_mask"          , causal_mask          , False, 0.5),
    ("none"        , None               , "causal_mask"          , causal_mask          , False, 2.0),
    # --- GQA + modifier + scale ---
    ("soft_cap"    , soft_cap           , "none"                 , None                 , True , 0.5),
    ("relative_pos", relative_positional, "causal_mask"          , causal_mask          , True , 1.0),
]
# fmt: on

PARAM_IDS = [
    f"score={score_name}_mask={mask_name}_gqa={gqa}_scale={scale}" for score_name, _, mask_name, _, gqa, scale in CASES
]


@pytest.mark.parametrize(
    ("score_mod_name", "score_mod_func", "mask_mod_name", "mask_mod_func", "use_gqa", "scale"),
    CASES,
    ids=PARAM_IDS,
)
def test_flexattention_minimum(
    score_mod_name,
    score_mod_func,
    mask_mod_name,
    mask_mod_func,
    use_gqa,
    scale,
    random_tensors,
    random_tensors_gqa,
):
    if use_gqa:
        q, k, v = random_tensors_gqa
    else:
        q, k, v = random_tensors
    validate_flexattention(score_mod_func, mask_mod_func, q, k, v, scale)


if __name__ == "__main__":
    pytest.main([__file__, "-v", "-s"])
```

</details>

cc @gramalingam @yuanyao-nv @justinchuby

gramalingam commented on onnx/defs/preview/defs.cc:
```diff
@@ -0,0 +1,509 @@
+    return static_cast<int32_t>(softmax_precision_attr->i());
+  }
+  // Auto-promote half-precision types to float32
+  if (input_elem_type == TensorProto::FLOAT16 || input_elem_type == TensorProto::BFLOAT16) {
```
Wouldn't this make even more sense for even smaller float-types (8bit, 4bit etc.)? The only place where it might not is if input type is double. Should we change the condition to if input_elem_type != Double?

yuanyao-nv commented on onnx/defs/preview/defs.cc:
```diff
@@ -0,0 +1,514 @@
+
+  // Compute attention scores by pre-scaling Q and K with sqrt(scale).
+  builder.Add("QF = Cast (Q)", "to", kFloat32)
+      .Add("KF = Cast (KAligned)", "to", kFloat32)
```
we did not do this casting for regular `Attention`. We did have a `softmax_precision` attribute which could lead to some casting around Softmax.

gramalingam commented on docs/docsgen/source/api/defs.md:
```diff
@@ -52,9 +52,11 @@ Domains officially supported in onnx package.
     from onnx.defs import (
         ONNX_DOMAIN,
         ONNX_ML_DOMAIN,
+        AI_ONNX_PREVIEW_DOMAIN,
```
Seems fine to me. (I mean the use "ai.onnx.preview")

gramalingam commented:
Overall, it looks fine for adding to preview. The preview domain is to enable experimentation. If we identify ways to improve this, we can address them later.

gramalingam commented:
Can you address the lint error? You can run "lintrunner -a"

mshr-h committed:
Replace assert with explicit RuntimeError for rank-4 input validation

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Fix unused parameter error

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Add real model score_mod test cases for FlexAttention

Add three new backend test cases demonstrating real-world score_mod
subgraph patterns:

- Causal mask (Qwen-3, Gemma-3, Llama-3 pattern): masks future tokens
  by setting scores to -inf where k_idx > q_idx
- Soft capping (Gemma-2 pattern): stabilizes attention scores via
  tanh(scores/cap) * cap
- Relative positional bias: adds (q_idx - k_idx) to attention scores

Each test includes the ONNX subgraph builder and expected output
computation for validation.

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Refactor FlexAttention computation to align K and V tensors and improve scaling factor handling

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Fix incorrect domain name in error message

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
address review comments

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Add BFLOAT16 support to FlexAttention softmax_precision

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
address review comments again

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
Add FlexAttention

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h commented:
Thank you for your review! Fixed lint errors and rebased from main branch. @gramalingam

mshr-h committed:
fix lint errors

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h committed:
update auto-gen files

Signed-off-by: Masahiro Hiramori <Hiramori.Masahiro@ct.MitsubishiElectric.co.jp>

mshr-h commented:
Can you retrigger the CI again? I forgot to add auto-gen files. Thank you for your help. @gramalingam

mshr-h commented:
Should we fix all the failure reports on SonarCloud Code Analysis? https://github.com/onnx/onnx/pull/7534/checks?check_run_id=72735135902

justinchuby commented:
Feel free to ignore them if they are not relevant. Thanks

mshr-h commented:
Thank you for your help. Please let me know if there's anything else I should address.
cc @gramalingam @justinchuby @ynimmaga

gramalingam commented:
It says one conversation needs to be marked as resolved ... before the merge happens automatically. I am trying to find it.
